src/daemon_helper.c

```diff
@@ -1652,36 +1749,166 @@ firmware_dir_walk(void)
  */
 static int fpga_state(void)
 {
-	FILE *fptr;
-	char buf[10];
-	char *state_operating = "operating";
-	char *state_unknown = "unknown";
+    const char *state_file_path = "/sys/class/fpga_manager/fpga0/state";
+    const char *state_operating = "operating";
+    const char *state_unknown = "unknown";
+    char read_buf[128];
+
+    if (read_single_line(state_file_path, read_buf, sizeof(read_buf)) < 0) {
+        DFX_ERR("Failed to determine the fpga state -"
+                " could not read state file");
+        return -1;
+    }
 
-	if (system("cat /sys/class/fpga_manager/fpga0/state >> state.txt")) {
-		DFX_ERR("Failed system() API");
-		return -1;
-	}
-	fptr = fopen("state.txt", "r");
-	if (fptr) {
-		if (fgets(buf, 10, fptr) == NULL) {
-			DFX_ERR("Failed to read fpga state");
-			buf[0] = 0;
-		}
-		fclose(fptr);
-		if (system("rm state.txt")) {
-			DFX_ERR("Failed system() API");
-		}
-		if ((strncmp(buf, state_operating, 9) == 0) || (strncmp(buf, state_unknown, 7) == 0))
-			return 0;
-		else
-			return -1;
-	}
+    DFX_DBG("FPGA state read as: `%s`", read_buf);
+
+    if (strcmp(read_buf, state_operating) == 0 ||
+        strcmp(read_buf, state_unknown) == 0) {
+        return 0;
+        }
+
+    DFX_ERR("FPGA is in a bad state. State: `%s`", read_buf);
+    return -1;
+}
+
+/**
+ * check_overlay_was_applied(...) - check that an overlay was applied
+ *
+ *
+ * @overlay_dir:      Path to the overlay directory in configfs.
+ * @requested_path:   The overlay path to write to the `path` attribute.
+ *
+ * This function checks that an overlay was properly applied by reading
+ * the overlay status and the overlay path by checking files in the configfs.
+ * It checks that "applied" is in the overlay's `status` attribute
+ * and that the `path` attribute contains the requested_path, i.e. the overlay
+ * file which was requested
+ *
+ * Return: 0 if the overlay status "applied" and path matches requested_path,
+ *        -1 on error or if either assertion is false
+ */
+static int check_overlay_was_applied(char *overlay_dir, char *requested_path)
+{
+    char full_path[256];
+    char read_buf[128];
+    const char *state_applied = "applied";
+
+    strip_trailing(overlay_dir, '/');
```

Comment:
we might not need this at all, it is best to verify behaviour for `//path` though

Reply:
I will test. usually works for bash cmds otherwise I will copy the buf before modifying

Reply:
I decided to make a copy into full_path. strip trailing and assemble the full path in 3 steps instead. This functionality is now moving into libdfx, however, so have made those changes there